src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncMethodInAsyncContext.cs

```diff
@@ -41,93 +42,162 @@ public sealed class UseAsyncMethodInAsyncContext : DiagnosticAnalyzer
         internal static DiagnosticDescriptor DescriptorNoAlternativeMethod = DiagnosticDescriptorHelper.Create(RuleId,
                                                                               s_localizableTitle,
                                                                               s_localizableMessageNoAlternative,
-                                                                              DiagnosticCategory.Reliability,
+                                                                              DiagnosticCategory.Performance,
                                                                               RuleLevel.IdeSuggestion,
                                                                               s_localizableDescription,
                                                                               isPortedFxCopRule: false,
                                                                               isDataflowRule: false);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor, DescriptorNoAlternativeMethod);
 
         public override void Initialize(AnalysisContext context)
         {
+
             context.EnableConcurrentExecution();
             context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
             context.RegisterOperationBlockStartAction(context =>
             {
-                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTask, out INamedTypeSymbol? taskType))
+                ConcurrentDictionary<string, string> qualifyingTypes = new();
+                qualifyingTypes.AddOrUpdate("Task", WellKnownTypeNames.SystemThreadingTasksTask, (k, v) => v);
+                qualifyingTypes.AddOrUpdate("TaskGeneric", WellKnownTypeNames.SystemThreadingTasksTask1, (k, v) => v);
+                qualifyingTypes.AddOrUpdate("ValueTask", WellKnownTypeNames.SystemThreadingTasksValueTask, (k, v) => v);
+                qualifyingTypes.AddOrUpdate("TaskAwaiter", WellKnownTypeNames.SystemRuntimeCompilerServicesTaskAwaiter, (k, v) => v);
+                qualifyingTypes.AddOrUpdate("ValueTaskAwaiter", WellKnownTypeNames.SystemRuntimeCompilerServicesValueTaskAwaiter, (k, v) => v);
+                qualifyingTypes.AddOrUpdate("IAsyncEnumberableGeneric", WellKnownTypeNames.SystemCollectionsGenericIAsyncEnumerable1, (k, v) => v);
+                qualifyingTypes.AddOrUpdate("AsyncMethodBuilderAttribute", WellKnownTypeNames.SystemRuntimeCompilerServicesAsyncMethodBuilderAttribute, (k, v) => v);
+
+                ConcurrentDictionary<string, ISymbol> syncBlockingTypes = GetSyncBlockingTypes(context.Compilation, qualifyingTypes);
```

Comment:
Any reason you're populating a `string-string` dictionary first and then using that to populate another `string-ISymbol` dictionary here? Wouldn't it be more efficient to just create a single `string-ISymbol` dictionary here and populate it? I'm thinking you can instantiate an empty `syncBlockingTypes` here first and then create a function like below:

```csharp
private void GetTypeAndAddToDictionary(string key, string typeName, ConcurrentDictionary<string, ISymbol> syncBlockingTypes, compilation)
{
    if (compilation.TryGetOrCreateTypeByMetadataName(typeName, out INamedTypeSymbol? typeValue))
                {
                    syncBlockingTypes.AddOrUpdate(typeName, typeValue, (k, v) => v);
                }
```

Then you can just call `GetTypeAndAddToDictionary` with `Task, TaskGeneric, ValueTask, etc`

You'll save on allocating and populating `qualifyingTypes` this way.